File: trees/Validate Binary Search Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class SolutionIsValidBST
{
    public class TreeNode
    {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode(int x) { val = x; }
    }

    private bool valid(TreeNode node, int min, int max)
    {
        if (node == null)
            return true;

        if (!(min < node.val && node.val < max))
            return false;

        return valid(node.left, min, node.val) && valid(node.right, node.val, max);
    }

    public bool IsValidBST(TreeNode root)
    {
        return valid(root, int.MinValue, int.MaxValue);
    }
}

File: trees/Validate Binary Search Tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isValidBST(root: Optional[TreeNode]) -> bool:

    def traverse(node, min_val, max_val):
        if not node:
            return True

        if not (min_val < node.val < max_val):
            return False

        return traverse(node.left, min_val, node.val) and traverse(node.right, node.val, max_val)

    return traverse(root, float("-inf"), float("inf"))
